src/app/components/CoinTable/index.tsx

```diff
@@ -9,137 +10,276 @@ import handleTableProgressBar from "@/app/utils/handleTableProgressBar";
-        let timer;
-        return () => {
-            if (!timer) {
-                func();
-                timer = setTimeout(() => {
-                    timer = null;
-                }, delay);
-            }
-        };
+  name: string;
+  id: string;
+  image: string;
+  symbol: string;
+  current_price: number;
+  total_volume: number;
+  market_cap: number;
+  circulating_supply: number;
+  total_supply: number;
+  price_change_percentage_1h_in_currency: number;
+  price_change_percentage_24h_in_currency: number;
+  price_change_percentage_7d_in_currency: number;
+}
+interface State {
+  currentCurrency: string;
+  coinTableList: any;
+}
+const currencySelect = (state: State) => state.currentCurrency;
+const coinTableListSelect = (state: State) => state.coinTableList.data;
+const coinsToDisplaySelect = (state: State) =>
+  state.coinTableList.coinsToDisplay;
+export default function CoinTable() {
+  const currentCurrency = useAppSelector(currencySelect);
+  const coinTableList = useAppSelector(coinTableListSelect);
+  const coinsToDisplay = useAppSelector(coinsToDisplaySelect);
+  const dispatch = useAppDispatch();
+  const handleThrottle = (func, delay: number) => {
+    let timer;
+    return () => {
+      if (!timer) {
+        func();
+        timer = setTimeout(() => {
+          timer = null;
+        }, delay);
+      }
     };
-    useEffect(() => {
-        const handleScroll = () => {
-            const { scrollTop,  offsetHeight} = document.documentElement;
-            if (window.innerHeight + scrollTop >= offsetHeight - 50) {
-                dispatch({type: "coinList/callCoins"});
-            }
-            return;
-        };
-        window.addEventListener("scroll" , handleThrottle(handleScroll, 2000));
-        return () => {
-            window.removeEventListener("scroll" , handleThrottle(handleScroll, 2000), true);
-        };
-      }, [dispatch]);
-    return (
-        <table className="w-full">
-            <tbody>
-            <tr>
-                <th>#</th>
-                <th>Name</th>
-                <th>Price</th>
-                <th>1hr%</th>
-                <th>24hr%</th>
-                <th>7d%</th>
-                <th>24hr volume / Market Cap</th>
-                <th>Circulating / Total Supply</th>
-                <th>Last 7d</th>
-            </tr>
+  };
+  useEffect(() => {
+    dispatch(fetchCoinTableList());
+  }, [coinsToDisplay]);
+  useEffect(() => {
+    const handleScroll = () => {
+      const { scrollTop, offsetHeight } = document.documentElement;
+      if (window.innerHeight + scrollTop >= offsetHeight - 50) {
+        dispatch({ type: "coinTableList/callCoins" });
+      }
+      return;
+    };
+    window.addEventListener("scroll", handleThrottle(handleScroll, 2000));
+    return () => {
+      window.removeEventListener(
+        "scroll",
+        handleThrottle(handleScroll, 2000),
+        true
+      );
+    };
+  }, []);
```

Comment:
extract this in a custom hook

Reply:
const tableRef = useRef()
useScroll(tableRef.current, () => {
   dispatch({ type: "coinTableList/callCoins" });
})

///

return (
    <table className="w-full" ref={tableRef}>




////



function useScroll(ref, cb) {
useEffect(() => {
    const handleScroll = () => {
      const { scrollTop, offsetHeight } = ref;
      if (window.innerHeight + scrollTop >= offsetHeight - 50) {
       cb()
      }
      return;
    };
    window.addEventListener("scroll", handleThrottle(handleScroll, 2000));
    return () => {
      window.removeEventListener(
        "scroll",
        handleThrottle(handleScroll, 2000),
        true
      );
    };
  }, []);
}